File: android/src/main/kotlin/dev/appspike/appspike_remote_config/AppspikeRemoteConfigPlugin.kt
package dev.appspike.appspike_remote_config

import android.content.Context
import android.util.Log
import dev.appspike.AppSpike
import dev.appspike.InitResult
import dev.appspike.remoteconfig.AppSpikeRemoteConfig
import dev.appspike.remoteconfig.ConfigUpdateListenerRegistration
import dev.appspike.remoteconfig.CustomSignals
import dev.appspike.remoteconfig.RemoteConfigFetchException
import dev.appspike.remoteconfig.RemoteConfigSettings
import dev.appspike.remoteconfig.RemoteConfigThrottledException
import dev.appspike.remoteconfig.ValueSource
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class AppspikeRemoteConfigPlugin :
    FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var applicationContext: Context

    private val remoteConfig = AppSpikeRemoteConfig

    private val mainScope = CoroutineScope(
        SupervisorJob() +
            Dispatchers.Main +
            CoroutineExceptionHandler { _, error ->
                Log.e(TAG, "Unhandled error in the Remote Config bridge", error)
            }
    )
    private var eventSink: EventChannel.EventSink? = null
    private var updateRegistration: ConfigUpdateListenerRegistration? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "dev.appspike/remote_config")
        channel.setMethodCallHandler(this)
        eventChannel =
            EventChannel(binding.binaryMessenger, "dev.appspike/remote_config/updates")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        updateRegistration?.remove()
        updateRegistration = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "ensureInitialized" -> launchGuarded(result) {
                remoteConfig.ensureInitialized()
                result.success(stateMap())
            }
            "setDefaults" -> {
                val defaults = call.argument<Map<String, Any?>>("defaults") ?: emptyMap()
                remoteConfig.setDefaults(defaults)
                result.success(stateMap())
            }
            "setConfigSettings" -> {
                val timeout = longArg(call, "fetchTimeoutSeconds") ?: 60L
                val interval = longArg(call, "minimumFetchIntervalSeconds") ?: 43_200L
                remoteConfig.setConfigSettings(
                    RemoteConfigSettings(
                        minimumFetchIntervalSeconds = interval,
                        fetchTimeoutSeconds = timeout,
                    )
                )
                result.success(stateMap())
            }
            "setCustomSignals" -> setCustomSignals(call, result)
            "fetch" -> fetch(call, result)
            "activate" -> launchGuarded(result) {
                val updated = remoteConfig.activate()
                result.success(stateMap(updated = updated))
            }
            "fetchAndActivate" -> fetchAndActivate(result)
            "getState" -> result.success(stateMap())
            "reset" -> {
                remoteConfig.reset()
                result.success(stateMap())
            }
            else -> result.notImplemented()
        }
    }

    private fun initialize(call: MethodCall, result: Result) {
        val apiKey = call.argument<String>("apiKey") ?: ""
        registerUpdateListener()
        AppSpike.initialize(
            context = applicationContext,
            apiKey = apiKey,
            modules = listOf(remoteConfig),
        ) { initResult ->
            launchGuarded(result) {
                when (initResult) {
                    is InitResult.Success -> result.success(mapOf("status" to "success"))
                    is InitResult.Error -> result.success(
                        mapOf("status" to "error", "message" to initResult.message)
                    )
                }
            }
        }
    }

    private fun setCustomSignals(call: MethodCall, result: Result) {
        val signals = call.argument<Map<String, Any?>>("signals") ?: emptyMap()
        val builder = CustomSignals.Builder()
        for ((key, value) in signals) {
            builder.put(key, value?.toString())
        }
        launchGuarded(result) {
            remoteConfig.setCustomSignals(builder.build())
            result.success(null)
        }
    }

    private fun fetch(call: MethodCall, result: Result) {
        val override = longArg(call, "minimumFetchIntervalSeconds")
        launchGuarded(result) {
            remoteConfig.fetch(override)
            result.success(stateMap())
        }
    }

    private fun fetchAndActivate(result: Result) {
        launchGuarded(result) {
            val updated = remoteConfig.fetchAndActivate()
            result.success(stateMap(updated = updated))
        }
    }

    private fun launchGuarded(result: Result, block: suspend () -> Unit) {
        mainScope.launch {
            try {
                block()
            } catch (error: Throwable) {
                fail(result, error)
            }
        }
    }

    private fun stateMap(updated: Boolean? = null): Map<String, Any?> {
        val values = HashMap<String, Any?>()
        for ((key, value) in remoteConfig.getAll()) {
            values[key] = mapOf(
                "value" to value.asString(),
                "source" to sourceName(value.getSource()),
            )
        }
        val info = remoteConfig.getInfo()
        val map = HashMap<String, Any?>()
        map["values"] = values
        map["lastFetchTimeMillis"] = info.lastFetchTimeMillis
        map["lastFetchStatus"] = info.lastFetchStatus.name
        map["fetchTimeoutSeconds"] = info.configSettings.fetchTimeoutSeconds
        map["minimumFetchIntervalSeconds"] =
            info.configSettings.minimumFetchIntervalSeconds
        if (updated != null) map["updated"] = updated
        return map
    }

    private fun sourceName(source: ValueSource): String = when (source) {
        ValueSource.REMOTE -> "remote"
        ValueSource.DEFAULT -> "default"
        ValueSource.STATIC -> "static"
    }

    private fun fail(result: Result, error: Throwable) {
        val details = if (error is RemoteConfigThrottledException) {
            mapOf("throttleEndTimeMillis" to error.throttleEndTimeMillis)
        } else {
            null
        }
        result.error(errorCode(error), error.message, details)
    }

    private fun errorCode(error: Throwable): String = when (error) {
        is RemoteConfigThrottledException -> "throttled"
        is RemoteConfigFetchException -> "fetch_failed"
        else -> "fetch_failed"
    }

    private fun longArg(call: MethodCall, name: String): Long? =
        when (val value = call.argument<Any?>(name)) {
            is Int -> value.toLong()
            is Long -> value
            is Number -> value.toLong()
            else -> null
        }

    private fun registerUpdateListener() {
        if (updateRegistration != null) return
        // Native 1.4.1 reshaped the listener to Firebase's two-method interface,
        // so the lambda overload now hands over a `ConfigUpdate` wrapper rather
        // than the bare key set.
        updateRegistration = remoteConfig.addOnConfigUpdateListener { update ->
            mainScope.launch {
                // No Result to fail here: a broken sink must not take the app
                // down, and the next activation will deliver its own keys.
                try {
                    eventSink?.success(update.updatedKeys.toList())
                } catch (error: Throwable) {
                    Log.e(TAG, "Failed to deliver a config update to Flutter", error)
                }
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private companion object {
        private const val TAG = "AppSpikeRemoteConfig"
    }
}

File: ios/Classes/AppspikeRemoteConfigPlugin.swift
import AppSpikeSDKCore
import AppSpikeSDKRemoteConfig
import Flutter
import Foundation

public class AppspikeRemoteConfigPlugin: NSObject, FlutterPlugin {
    private let remoteConfig = AppSpikeRemoteConfig.shared
    private var updatesSink: FlutterEventSink?
    private var updateRegistration: ConfigUpdateListenerRegistration?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AppspikeRemoteConfigPlugin()
        let channel = FlutterMethodChannel(
            name: "dev.appspike/remote_config",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: channel)

        let events = FlutterEventChannel(
            name: "dev.appspike/remote_config/updates",
            binaryMessenger: registrar.messenger()
        )
        events.setStreamHandler(instance)
    }

    public func handle(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        let arguments = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "initialize":
            initialize(arguments, result)
        case "ensureInitialized":
            // No explicit await hook is needed; onSessionReady has already
            // built the core by the time initialize's callback fired.
            result(self.stateMap())
        case "setDefaults":
            let defaults = arguments["defaults"] as? [String: Any] ?? [:]
            remoteConfig.setDefaults(defaults)
            result(self.stateMap())
        case "setConfigSettings":
            let timeout = doubleValue(arguments["fetchTimeoutSeconds"]) ?? 60
            let interval =
                doubleValue(arguments["minimumFetchIntervalSeconds"]) ?? 43_200
            remoteConfig.configSettings = RemoteConfigSettings(
                minimumFetchInterval: interval,
                fetchTimeout: timeout
            )
            result(self.stateMap())
        case "setCustomSignals":
            setCustomSignals(arguments, result)
        case "fetch":
            fetch(arguments, result)
        case "activate":
            activate(result)
        case "fetchAndActivate":
            fetchAndActivate(result)
        case "getState":
            result(self.stateMap())
        case "reset":
            remoteConfig.reset()
            result(self.stateMap())
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method handlers

    private func initialize(
        _ arguments: [String: Any],
        _ result: @escaping FlutterResult
    ) {
        let apiKey = arguments["apiKey"] as? String ?? ""
        registerUpdateListener()
        AppSpike.shared.initialize(
            apiKey: apiKey,
            modules: [remoteConfig]
        ) { initResult in
            DispatchQueue.main.async {
                switch initResult {
                case .success:
                    result(["status": "success"])
                case .error(let message):
                    result(["status": "error", "message": message])
                }
            }
        }
    }

    private func setCustomSignals(
        _ arguments: [String: Any],
        _ result: @escaping FlutterResult
    ) {
        let signals = arguments["signals"] as? [String: Any] ?? [:]
        let builder = CustomSignals.Builder()
        for (key, value) in signals {
            builder.put(key: key, value: value is NSNull ? nil : "\(value)")
        }
        let customSignals = builder.build()
        Task {
            do {
                try await remoteConfig.setCustomSignals(customSignals)
                DispatchQueue.main.async { result(nil) }
            } catch {
                DispatchQueue.main.async {
                    result(self.fetchError(error))
                }
            }
        }
    }

    private func fetch(
        _ arguments: [String: Any],
        _ result: @escaping FlutterResult
    ) {
        let override = doubleValue(arguments["minimumFetchIntervalSeconds"])
        Task {
            do {
                if let override {
                    try await remoteConfig.fetch(
                        withExpirationDuration: override
                    )
                } else {
                    // Native 1.4.1 exposes a distinct no-argument `fetch()`
                    // that honors `configSettings.minimumFetchInterval`
                    // (Firebase parity), so the interval no longer has to be
                    // read out and passed back in.
                    try await remoteConfig.fetch()
                }
                DispatchQueue.main.async { result(self.stateMap()) }
            } catch {
                DispatchQueue.main.async {
                    result(self.fetchError(error))
                }
            }
        }
    }

    private func activate(_ result: @escaping FlutterResult) {
        Task {
            let updated = await remoteConfig.activate()
            DispatchQueue.main.async {
                result(self.stateMap(updated: updated))
            }
        }
    }

    private func fetchAndActivate(_ result: @escaping FlutterResult) {
        Task {
            do {
                let status = try await remoteConfig.fetchAndActivate()
                let updated = status == .successFetchedFromRemote
                DispatchQueue.main.async {
                    result(self.stateMap(updated: updated))
                }
            } catch {
                DispatchQueue.main.async {
                    result(self.fetchError(error))
                }
            }
        }
    }

    // MARK: - Serialization

    private func stateMap(updated: Bool? = nil) -> [String: Any] {
        var values: [String: Any] = [:]
        for (key, value) in remoteConfig.allConfigValues() {
            values[key] = [
                "value": value.stringValue,
                "source": sourceName(value.source),
            ]
        }
        var map: [String: Any] = [
            "values": values,
            "lastFetchTimeMillis":
                Int(
                    (remoteConfig.lastFetchTime?.timeIntervalSince1970 ?? 0)
                        * 1000
                ),
            "lastFetchStatus": statusName(remoteConfig.lastFetchStatus),
            "fetchTimeoutSeconds":
                Int(remoteConfig.configSettings.fetchTimeout),
            "minimumFetchIntervalSeconds":
                Int(remoteConfig.configSettings.minimumFetchInterval),
        ]
        if let updated {
            map["updated"] = updated
        }
        return map
    }

    private func statusName(_ status: RemoteConfigFetchStatus) -> String {
        switch status {
        case .success: return "SUCCESS"
        case .failure: return "FAILURE"
        case .throttled: return "THROTTLED"
        case .noFetchYet: return "NO_FETCH_YET"
        @unknown default: return "NO_FETCH_YET"
        }
    }

    private func sourceName(_ source: RemoteConfigSource) -> String {
        switch source {
        case .remote: return "remote"
        case .default: return "default"
        case .static: return "static"
        }
    }

    private func throttleEndTimeMillis(
        _ error: RemoteConfigThrottledError
    ) -> Int64 {
        return Int64(error.throttleEndTimeInterval * 1000)
    }

    private func fetchError(_ error: Error) -> FlutterError {
        if let throttled = error as? RemoteConfigThrottledError {
            return FlutterError(
                code: "throttled",
                message: throttled.message,
                details: [
                    "throttleEndTimeMillis":
                        NSNumber(value: throttleEndTimeMillis(throttled))
                ]
            )
        }
        let message =
            (error as? RemoteConfigFetchError)?.message
            ?? error.localizedDescription
        return FlutterError(
            code: "fetch_failed",
            message: message,
            details: nil
        )
    }

    private func doubleValue(_ value: Any?) -> Double? {
        if let number = value as? NSNumber { return number.doubleValue }
        return nil
    }

    // MARK: - Config update listener

    private func registerUpdateListener() {
        guard updateRegistration == nil else { return }
        // Native 1.4.1 reshaped this to Firebase's two-argument completion
        // `(RemoteConfigUpdate?, Error?)`, under the
        // `remoteConfigUpdateCompletion:` label. The error arm never fires
        // in this version — a failing fetch throws to the `fetch` caller
        // instead — and the Flutter event channel carries changed keys only,
        // so a nil update is simply not forwarded.
        updateRegistration = remoteConfig.addOnConfigUpdateListener(
            remoteConfigUpdateCompletion: { [weak self] update, _ in
                guard let update else { return }
                DispatchQueue.main.async {
                    self?.updatesSink?(Array(update.updatedKeys))
                }
            }
        )
    }
}

extension AppspikeRemoteConfigPlugin: FlutterStreamHandler {
    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        updatesSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        updatesSink = nil
        return nil
    }
}
